File: solutions/python/resistor-color-duo/1/resistor_color_duo.py
COLOR_BY_VALUE = [
    "black",
    "brown",
    "red",
    "orange",
    "yellow",
    "green",
    "blue",
    "violet",
    "grey",
    "white",
]

def value(colors):
    return COLOR_BY_VALUE.index(colors[0]) * 10 + COLOR_BY_VALUE.index(colors[1])

File: solutions/prolog/resistor-color-duo/1/resistor_color_duo.pl
value(Colors, Value) :-
    [First, Second | Rest] = Colors,
    maplist(color_code, [First, Second], Digits),
    reverse(Digits, ReversedDigits),
    combined_value(ReversedDigits, Value).

combined_value([], 0).
combined_value([LeastDigit | Tail], Value) :-
    combined_value(Tail, TailValue),
    Value is TailValue * 10 + LeastDigit.

color_code(Color, Code) :-
    colors(AllColors),
    nth0(Code, AllColors, Color).

colors(Colors) :-
    Colors = [
        "black", "brown", "red", "orange", "yellow", 
        "green", "blue", "violet", "grey", "white"
    ].
